<a href="https://colab.research.google.com/github/raj-vijay/ml/blob/master/06.Dimensionality%20Reduction/06_LASSO_Regressor_Feature_Selection_ANSUR_Dataset.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

In [ ]:
ansur_df = pd.read_csv('ANSUR.csv')

In [ ]:
ansur_df.head()

Branch     Component  ...   BMI_class  Height_class
0          Combat Support  Regular Army  ...  Overweight        Normal
1  Combat Service Support  Regular Army  ...      Normal        Normal
2  Combat Service Support  Regular Army  ...      Normal          Tall
3  Combat Service Support  Regular Army  ...  Overweight        Normal
4             Combat Arms  Regular Army  ...  Overweight        Normal

[5 rows x 99 columns]

In [ ]:
ansur_df = ansur_df.drop(['Branch',	'Gender','Component', 'BMI_class',	'Height_class'], axis = 1)

In [ ]:
# list = ['elbowrestheight', 'wristcircumference', 'anklecircumference', 'buttockheight', 'crotchheight']

In [ ]:
# df = ansur_df[list]

In [ ]:
X = ansur_df.drop('BMI', axis = 1)

In [ ]:
y = ansur_df['BMI']

In [ ]:
from sklearn.linear_model import LogisticRegression, Lasso
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.feature_selection import RFE
from sklearn.ensemble import RandomForestClassifier

In [ ]:
scaler = StandardScaler()

In [ ]:
# Set the test size to 30% to get a 70-30% train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)

# Fit the scaler on the training features and transform these in one go
X_train_std = scaler.fit_transform(X_train)

# Create the Lasso model 
la = Lasso()

# Fit it to the standardized training data
la.fit(X_train_std, y_train)

Lasso(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=1000,
      normalize=False, positive=False, precompute=False, random_state=None,
      selection='cyclic', tol=0.0001, warm_start=False)

**Lasso model results**

After training the Lasso model, we score its predictive capacity (R2) on the test set and count how many features are ignored because their coefficient is reduced to zero.

In [ ]:
# Transform the test set with the pre-fitted scaler
X_test_std = scaler.transform(X_test)

# Calculate the coefficient of determination (R squared) on X_test_std
r_squared = la.score(X_test_std, y_test)
print("The model can predict {0:.1%} of the variance in the test set.".format(r_squared))

# Create a list that has True values when coefficients equal 0
zero_coef = la.coef_ == 0

# Calculate how many features have a zero coefficient
n_ignored = sum(zero_coef)
print("The model has ignored {} out of {} features.".format(n_ignored, len(la.coef_)))

The model can predict 83.9% of the variance in the test set.
The model has ignored 83 out of 93 features.


**Adjusting the regularization strength**

The current Lasso model has an R2 score of 83.9%. 

When a model applies overly powerful regularization it can suffer from high bias, hurting its predictive power.

Thus we improve the balance between predictive power and model simplicity by tweaking the alpha parameter.

In [ ]:
# Find the highest alpha value with R-squared above 98%
la = Lasso(alpha=0.1, random_state=0)

# Fits the model and calculates performance stats
la.fit(X_train_std, y_train)
r_squared = la.score(X_test_std, y_test)
n_ignored_features = sum(la.coef_ == 0)

# Print peformance stats 
print("The model can predict {0:.1%} of the variance in the test set.".format(r_squared))
print("{} out of {} features were ignored.".format(n_ignored_features, len(la.coef_)))

The model can predict 98.6% of the variance in the test set.
72 out of 93 features were ignored.
